### PR TITLE
check build exists before make actual build

### PR DESCRIPTION
**Pull request by krzyzacy**
how about this?
tested locally and seems to work

/assign @ixdy @BenTheElder

### CODE REVIEW

**Comment by krzyzacy** [comment]
@BenTheElder suggested we can do `bazel --batch build //:save_git_version` and get the version from `bazel-genfiles/version`, downside is that we need to use a bazel image now for none bazel build.

@ixdy WDYT

**Comment by BenTheElder** [comment]
> downside is that we need to use a bazel image now for none bazel build.

using kubekins is fine, basically a no-op, but making the non-bazel CI job use bazel is kinda 🤷‍♂️ 
OTOH the bazel build actually exposes a version target pretty cleanly, make / shell stuff doesn't really.

File: scenarios/kubernetes_build.py [hunk 1]
```diff
@@ -30,8 +30,43 @@ def check(*cmd):
...
+    """ check if a k8s/federation build with same version
+        already exists in remote path
+    """
+    if not os.path.exists('hack/lib/version.sh'):
+        return False
+
+    # gross, but, /shrug
+    gross = '. hack/lib/version.sh && KUBE_ROOT=. ' \
+            'kube::version::get_version_vars && echo "${KUBE_GIT_VERSION-}"'
+    version = ''
```
Comment (BenTheElder) [inline scenarios/kubernetes_build.py]: discussing below but this is indeed gross :-(

**Comment by krzyzacy** [comment]
(updated to use bazel target)

**Comment by krzyzacy** [comment]
PTAL, I'd like to unbreak kops

File: scenarios/kubernetes_build.py [hunk 2]
```diff
@@ -30,8 +30,40 @@ def check(*cmd):
...
+    print >>sys.stderr, 'Run:', cmd
+    return subprocess.check_output(cmd)
+
+def check_build_exists(gcs):
+    """ check if a k8s/federation build with same version
+        already exists in remote path
+    """
+    version = ''
+    try:
+        check('bazel', '--batch', 'build', '//:save_git_version')
```
Comment (ixdy) [inline scenarios/kubernetes_build.py]: a slightly better way (which is similar):
```shell
hack/print-workspace-status.sh | grep gitVersion | awk '{print $2}'
```

Comment (BenTheElder) [reply scenarios/kubernetes_build.py]: Definitely use this :+1: :-)

Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: this is equivalent hacky-ish...

we probably should add a make target to print the version in k/k...

Comment (BenTheElder) [reply scenarios/kubernetes_build.py]: This isn't very hacky, print-workspace-status.sh spits out the gitVersion (and other things) and then we're just grabbing the actual version. 

We'd have to backport a make target to older releases

File: scenarios/kubernetes_build.py [hunk 3]
```diff
@@ -30,8 +30,40 @@ def check(*cmd):
...
+                version = fp.read()
+    except subprocess.CalledProcessError as exc:
+        # fallback with doing a real build
+        print >>sys.stderr, 'Failed to get k8s version, continue: %s' % exc
+        return False
+
+    if version:
+        if not gcs:
+            gcs = 'gs://kubernetes-release-dev'
+        gcs = os.path.join(gcs, 'ci', version)
```
Comment (ixdy) [inline scenarios/kubernetes_build.py]: need to add the suffix (if set) to `ci` too.

File: scenarios/kubernetes_build.py [hunk 4]
```diff
@@ -41,6 +73,13 @@ def main(args):
         print >>sys.stderr, (
             'Scenario should only run from either kubernetes or federation directory!')
         sys.exit(1)
+
+    # pre-check if target build exists in gcs bucket or not
+    # if so, don't make duplicated builds
+    if check_build_exists(args.release):
+        print >>sys.stderr, 'build already exists, exit'
+        sys.exit(1)
```
Comment (ixdy) [inline scenarios/kubernetes_build.py]: do we want to fail builds? isn't that exactly the same problem we had before we allowed duplicate uploads?

Comment (BenTheElder) [reply scenarios/kubernetes_build.py]: This should definitely be `sys.exit(0)`

Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: good catch!

Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: done

File: scenarios/kubernetes_build.py (hunk 3 above)
Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: done

File: scenarios/kubernetes_build.py (hunk 2 above)
Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: done

**Comment by krzyzacy** [comment]
seems to work (my laptop k8s branch is still on 1.4 🤦‍♂️ )
```
../test-infra/scenarios/kubernetes_build.py 
Run: ('hack/print-workspace-status.sh',)
Run: ('gsutil', 'ls', 'gs://kubernetes-release-dev/ci/v1.4.0-alpha.2.26172+054490609f57c5')
CommandException: One or more URLs matched no objects.
gcs path gs://kubernetes-release-dev/ci/v1.4.0-alpha.2.26172+054490609f57c5 does not exist yet, continue
```

File: scenarios/kubernetes_build.py [hunk 5]
```diff
@@ -30,8 +31,49 @@ def check(*cmd):
...
+
+    if version:
+        if not gcs:
+            gcs = 'gs://kubernetes-release-dev'
+        mode = 'ci'
+        if suffix:
+            mode = os.path.join(mode, suffix)
+        gcs = os.path.join(gcs, mode, version)
+        try:
+            check('gsutil', 'ls', gcs)
```
Comment (ixdy) [inline scenarios/kubernetes_build.py]: I wonder if we should suppress stderr. I foresee
```
CommandException: One or more URLs matched no objects.
```
potentially confusing users.

**Comment by ixdy** [comment]
other than minor suggestion this lgtm

File: scenarios/kubernetes_build.py (hunk 5 above)
Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: do we have a standard way to suppress stderr?

Comment (ixdy) [reply scenarios/kubernetes_build.py]: something like https://stackoverflow.com/a/11269627, probably?

Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: yeah, that works

```
../test-infra/scenarios/kubernetes_build.py --suffix=boo
Run: ('hack/print-workspace-status.sh',)
Run: ('gsutil', 'ls', 'gs://kubernetes-release-dev/ci/boo/v1.4.0-alpha.2.26172+054490609f57c5')
gcs path gs://kubernetes-release-dev/ci/boo/v1.4.0-alpha.2.26172+054490609f57c5 does not exist yet, continue
Run: ('make', 'clean')
...
```

File: scenarios/kubernetes_build.py [hunk 6]
```diff
@@ -30,8 +31,55 @@ def check(*cmd):
     print >>sys.stderr, 'Run:', cmd
     subprocess.check_call(cmd)
 
+def check_no_stdout(*cmd):
+    """Log and run the command, suppress stdout, raising on errors."""
```
Comment (ixdy) [inline scenarios/kubernetes_build.py]: I think we want to suppress stderr? otherwise we'll still get the misleading "not found" error message.

Comment (krzyzacy) [reply scenarios/kubernetes_build.py]: done for both now

**Comment by ixdy** [comment]
/lgtm
/hold

to merge on a Friday evening or not?

**Comment by krzyzacy** [comment]
😂 I'll merge a bit later when I can 👀 it 🔥

**Comment by BenTheElder** [comment]
BTW: this doesn't cover cross IIRC ?

**Comment by krzyzacy** [comment]
@BenTheElder why?

**Comment by BenTheElder** [comment]
oh nevermind, I could have sworn that job didn't use the build scenario, the presubmit uses execute.

/shrug
/hold cancel

**Comment by BenTheElder** [comment]
Friday night is the best time to change the build. Will keep 👀 out

**Comment by BenTheElder** [comment]
So far seems fine:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-build/8375
http://gcsweb.k8s.io/gcs/kubernetes-release-dev/ci/v1.10.0-alpha.3.496+4f76fc20420f00

**Comment by ixdy** [comment]
> Friday night is the best time to change the build.

/yolo

**Pull request merged**